xarray/conventions.py

```diff
@@ -139,6 +139,15 @@ def _infer_dtype(array, name=None):
     if array.size == 0:
         return np.dtype(float)
 
+    native_dtypes = set(map(lambda x: type(x), array.flatten()))
```

Comment:
I am afraid that this can be very slow but I don't have a good idea either...

```suggestion
    native_dtypes = set(np.vectorize(type, otypes=[object])(a.ravel()))
```
should be a bit faster, but probably does not fare very well for _really big arrays_<sup>:tm:</sup>, either

(n.b. `ravel` makes a view while `flatten` creates a copy)